File: 287-find-the-duplicate-number/find-the-duplicate-number.java
class Solution {

    public int findDuplicate(int[] nums) {
        
        // VISITED IDX

        /*
        int len = nums.length;
        for(int num : nums) {
            int idx = Math.abs(num);
            if(nums[idx] < 0) {
                return idx;
            }

            nums[idx] = -nums[idx];
        }

        return len;
        */

        // SLOW FAST POINTER

        int slow = 0;
        int fast = 0;

        do{
            slow = nums[slow];
            fast = nums[nums[fast]];
        } while(slow != fast);

        slow = 0;

        while(slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }

        return slow;
    }
}

File: 287-find-the-duplicate-number/find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {

        // Circular Linked List Concept with Floyd's Algorithm
        // T.C => O(N) && S.C => (1)
        
        int slow = nums[0];
        int fast = nums[nums[0]];

        while(slow != fast) {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }

        fast = 0;
        while(slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }

        return fast;
    }
};